cmd/kubectl-gadget/utils/trace.go

```diff
@@ -465,52 +468,77 @@ func getTraceWatcher(traceID string) (watch.Interface, error) {
 func waitForCondition(traceID string, conditionFunction func(*gadgetv1alpha1.Trace) bool) (*gadgetv1alpha1.TraceList, error) {
 	var returnedTraces gadgetv1alpha1.TraceList
 
-	tracesNumber := 0
-	watchedTracesNumber := 0
+	traceList, err := getTraceListFromID(traceID)
+	if err != nil {
+		return nil, err
+	}
+
+	tracesNumber := len(traceList.Items)
+	tracesReady := 0
+
+	// Maybe some traces are already ready?
+	for _, trace := range traceList.Items {
+		if conditionFunction(&trace) {
+			tracesReady++
+		}
+	}
 
+	// If they are all ready we can quit now without having to watch events.
+	if tracesReady == tracesNumber {
+		return traceList, nil
+	}
+
+	// Otherwise, we will need to watch events on them.
+	// For this, we will get a watcher on all the traces which share the same ID.
 	// Get a watcher on all the traces which have the same ID.
 	// Indeed, all the traces on different nodes but linked to one gadget share
 	// the same ID.
-	watcher, err := getTraceWatcher(traceID)
+	// We will also begin to monitor events since the above GET of the traces
+	// list.
+	watcher, err := getTraceWatcher(traceID, traceList.ListMeta.ResourceVersion)
 	if err != nil {
 		return nil, err
 	}
 
 	nodeErrors := make(map[string]string)
 	nodeWarnings := make(map[string]string)
 
+	// We will only watch the traces which were not ready.
+	tracesToWatch := tracesNumber - tracesReady
+	watchedTracesNumber := 0
+
```

Comment:
As described in the comment above. If you do this, you have to discard the event coming from the already read trace. Otherwise, there is the risk that the same trace that is already ready make the watcher condition goes out and we do not wait for the other traces to be ready.

To solve this, I propose to create a map where you store for each trace its status (whether it satisfies the condition or not) and then you return only when all the traces are ready. What do you think?

Reply:
> As described in the comment above. If you do this, you have to discard the event coming from the already read trace. Otherwise, there is the risk that the same trace that is already ready make the watcher condition goes out and we do not wait for the other traces to be ready.

This should never happens due to using the list `ResourceVersion` (RV).
Indeed, by using this RV it would prevent receiving the events which modified the traces to be ready as those events belong to the past.

Reply:
The problem is that when the trace-controller manages an operation it modifies the trace twice: the first one to remove the operation from the trace annotations (generating one event on the CLI) and the second one to update the status with the new state or the output (generating another event). 

Therefore, if the event that modified the trace to be ready is the one for the change on the annotations, then the watcher will receive another event for the change in the status (this is the one that needs to be discarded). Similarly, if the trace is not ready when the watcher starts, it could receive two consecutive events from the same trace and wrongly decide that all traces are ready (again here, the second one needs to be discarded). BTW, I reproduced the second case.

That's why I propose to have a map to avoid increasing the `watchedTracesNumber` twice for the same trace. What do you think?

Reply:
I see.
Actually the code would not handle if, for example, the same operation is called twice.
A simple counter does, sadly, not suffice.
I will add a map of traces we watched, if a trace is already watched, we would not watch it again.